### PR TITLE
#1833 Shortcuts Lab: Export slide as image

### LINKED ISSUE #1833: Shortcuts Lab Feature: Export Slide as Image
As mentioned by Prof @damithc in #195, we will add a context menu option when user right clicks on the slide/slide thumbnail that allows user to export the slide as an image directly on their computer. 

Currently, the only way to do so in PowerPoint is to use "Save As" and select the file type to be an image file type. 

This context menu option will be added as a new feature under Shortcuts Lab.

### PR DESCRIPTION
**Pull request by blewjy**
Fixes #1833 

This PR adds a new feature to ShortcutsLab, which allows user to export a slide (or multiple slides) as images via the context menu.

export-slide-as-image-ss 0

**Behaviour:**

- If only one slide is selected, slide will be exported as a PNG image and saved in the location specified by the user via a Save File Dialog.
- If multiple slides are selected, slides will be saved in a new folder with the folder name specified by the user.

### CODE REVIEW

File: PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageActionHandler.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            System.Windows.Forms.SaveFileDialog saveFileDialog = new System.Windows.Forms.SaveFileDialog();
+            saveFileDialog.Title = "Export Slide As Image";
+            saveFileDialog.DefaultExt = "png";
+            saveFileDialog.Filter = "Images|*.png;*.bmp;*.jpg";
+
+            System.Windows.Forms.DialogResult result = saveFileDialog.ShowDialog();
+
+            if (result == System.Windows.Forms.DialogResult.OK)
+            {
+
```
Comment (ChesterSng) [inline PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageActionHandler.cs]: Blank lines :P

File: PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageEnabledHandler.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using PowerPointLabs.ActionFramework.Common.Attribute;
+using PowerPointLabs.ActionFramework.Common.Extension;
+using PowerPointLabs.ActionFramework.Common.Interface;
+using PowerPointLabs.TextCollection;
+using PowerPointLabs.Utils;
+
+namespace PowerPointLabs.ActionFramework.ShortcutsLab
```
Comment (ChesterSng) [inline PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageEnabledHandler.cs]: Doesn't seem like this class is needed

**Comment by blewjy** [comment]
@ChesterSng yes this is a behaviour that I was also considering how best to implement. It is absolutely possible to save all the images to the specified folder, however if there are many slides it may unnecessarily clutter the selected folder.

I decided to go with this implementation in the end because this is how PowerPoint does it as well. If you have multiple slides in your presentation and you click "Save As", and select one of the image file type extensions, PowerPoint will prompt if you want to save the current slide only or all slides. If you select all slides, PowerPoint will save it into a new folder with the name that you specified, and each image will have the corresponding Slide's name.

**Comment by ChesterSng** [comment]
Sounds good! @blewjy 

Just a minor nit here: 
image

The selected slides ~will~ are saved in the folder...

"Will" sounds like more like asking for permission

Otherwise, LGTM!

File: PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageEnabledHandler.cs (hunk 2 above)
Comment (blewjy) [reply PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageEnabledHandler.cs]: RemoveD!

File: PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageActionHandler.cs (hunk 1 above)
Comment (blewjy) [reply PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/ExportSlideAsImage/ExportSlideAsImageActionHandler.cs]: Removed!

**Comment by blewjy** [comment]
@ChesterSng made the changes! Thanks for your review

File: PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs [hunk 3]
```diff
@@ -191,6 +192,51 @@ private static void DeleteSpecificFilePath(string filePath)
...
+                    {
+                        string fileName = folderName + "\\" + slide.Name + ".png";
+                        ExportSlide(slide, fileName, magnifyRatio);
+                    }
+                }
+                catch (Exception)
+                {
+
+                    // Failed to create directory, we save the images all to the specified path
+                    foreach (Slide slide in slides)
```
Comment (YuPeiHenry) [inline PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs]: Will there be a message displayed for failure to create directory?

File: PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs [hunk 4]
```diff
@@ -191,6 +192,51 @@ private static void DeleteSpecificFilePath(string filePath)
...
+            {
+                return;
+            } 
+            else if (slides.Count == 1)
+            {
+                ExportSlide(slides[0], exportPath, magnifyRatio);
+                return;
+            }
+
+            if (!Directory.Exists(exportPath))
```
Comment (YuPeiHenry) [inline PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs]: Change this into a guard clause? Seems like there is no code outside this if block.

File: PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs (hunk 3 above)
Comment (YuPeiHenry) [inline PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs]: Will there be a message displayed for failure to create directory?

File: PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs (hunk 4 above)
Comment (blewjy) [reply PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs]: I refactored this to remove the if check.

File: PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs (hunk 3 above)
Comment (blewjy) [reply PowerPointLabs/PowerPointLabs/Utils/GraphicsUtil.cs]: Added a message box!

**Comment by blewjy** [comment]
@YuPeiHenry thanks for your review! I tweaked the code a little to remove the if check, and also enabled the exported images to be saved in a folder of some default name in the event that the user entered a name that is already taken. 

***Note:*** The `Save File Dialog` actually handles this already, that is, if the user were to enter a name that is already taken by another file (this is when exporting a single slide), user will be prompted if he wants to overwrite the file. It is possible, however, for the user to enter a folder name (when exporting multiple slides) that is already taken, and append with some extension (for e.g.: `.png`), and the save file dialog does not catch this. Our code will then catch this and create a folder with some default name instead.

**Pull request merged**